File: leetcode/algorithms/hash-table/single-number-iii.js
/*
    leetcode
    category: "Algorithms"
    problem title: "Single Number III"
*/

/**
 * @param {number[]} nums
 * @return {number[]}
 */
var singleNumber = function(nums) {
    
    return [...nums.reduce((numSet,num) => {
        if(numSet.has(num)){
            numSet.delete(num);
        } else {
            numSet.add(num);
        }
        return numSet;
    },new Set())]
    
};

File: leetcode/algorithms/hash-table/single-number-iii.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Single Number III"
*/

class Solution {
    public int[] singleNumber(int[] nums) {
        HashSet<Integer> numbersSet = new HashSet<Integer>();
        int[] answer = new int[2];
        for(int num:nums){
            if(numbersSet.contains(num)){
                numbersSet.remove(num);
            } else {
                numbersSet.add(num);
            }
        }
        int i=0;
        for(Integer ans : numbersSet){
            answer[i]=ans.intValue();
            i++;
        }
        return answer;
    }
}